packages/pdfkit/src/reference.js

```diff
@@ -1,91 +1,74 @@
 import PDFObject from './object';
 
-class PDFReference extends stream.Writable {
-  constructor(document, id, data) {
-    super({ decodeStrings: false });
-
-    this.finalize = this.finalize.bind(this);
+class PDFReference {
+  constructor(document, id, data = {}) {
     this.document = document;
     this.id = id;
-    if (data == null) {
-      data = {};
-    }
     this.data = data;
-
     this.gen = 0;
-    this.deflate = null;
     this.compress = this.document.compress && !this.data.Filter;
     this.uncompressedLength = 0;
-    this.chunks = [];
+    this.buffer = [];
   }
 
-  initDeflate() {
-    this.data.Filter = 'FlateDecode';
-
-    this.deflate = zlib.createDeflate();
-    this.deflate.on('data', (chunk) => {
-      this.chunks.push(chunk);
-      return (this.data.Length += chunk.length);
-    });
-
-    return this.deflate.on('end', this.finalize);
-  }
-
-  _write(chunk, encoding, callback) {
-    if (!(chunk instanceof Uint8Array)) {
+  write(chunk) {
+    if (!Buffer.isBuffer(chunk)) {
       chunk = Buffer.from(chunk + '\n', 'binary');
     }
 
     this.uncompressedLength += chunk.length;
     if (this.data.Length == null) {
       this.data.Length = 0;
     }
-
+    this.buffer.push(chunk);
+    this.data.Length += chunk.length;
     if (this.compress) {
-      if (!this.deflate) {
-        this.initDeflate();
-      }
-      this.deflate.write(chunk);
-    } else {
-      this.chunks.push(chunk);
-      this.data.Length += chunk.length;
+      return (this.data.Filter = 'FlateDecode');
     }
-
-    return callback();
   }
 
-  end() {
-    super.end(...arguments);
-
-    if (this.deflate) {
-      return this.deflate.end();
+  end(chunk) {
+    if (chunk) {
+      this.write(chunk);
     }
-
     return this.finalize();
   }
 
   finalize() {
     this.offset = this.document._offset;
 
+    const encryptFn = this.document._security
+      ? this.document._security.getEncryptFn(this.id, this.gen)
+      : null;
+
+    if (this.buffer.length) {
+      this.buffer = Buffer.concat(this.buffer);
+      if (this.compress) {
+        this.buffer = zlib.deflateSync(this.buffer);
+      }
+
+      if (encryptFn) {
+        this.buffer = encryptFn(this.buffer);
+      }
+
+      this.data.Length = this.buffer.length;
+    }
+
     this.document._write(`${this.id} ${this.gen} obj`);
-    this.document._write(PDFObject.convert(this.data));
+    this.document._write(PDFObject.convert(this.data, encryptFn));
 
-    if (this.chunks.length) {
+    if (this.buffer.length) {
       this.document._write('stream');
-      for (let chunk of Array.from(this.chunks)) {
-        this.document._write(chunk);
-      }
+      this.document._write(this.buffer);
 
-      this.chunks.length = 0; // free up memory
+      this.buffer = []; // free up memory
       this.document._write('\nendstream');
     }
 
     this.document._write('endobj');
-    return this.document._refEnd(this);
+    this.document._refEnd(this);
   }
-
```

Comment:
Nit: we probably want to keep this space

Reply:
Makes sense, I missed this during merge - I've restored it 🙇